### PR TITLE
NIFI-11443 Route Python Framework Logging to SLF4J

### PR DESCRIPTION
**Pull request by exceptionfactory**
# Summary

[NIFI-11443](https://issues.apache.org/jira/browse/NIFI-11443) Updates Python Process log handling to route messages to SLF4J so that logging levels and appenders can be managed using Logback.

These changes remove the need for separate Python log settings in `nifi.properties` and support translating Python log levels and logger names to Logback.

The implementation changes the Python logging format to use the Python [level numbers](https://docs.python.org/3/library/logging.html#levels) for optimized parsing. The implementation changes Python logging to write messages to the standard output stream, and the `PythonProcess` builder creates a Java Virtual Thread to read lines from the `BufferedReader` containing process output stream messages.

These changes eliminate the dedicated `nifi-python.log` and enable Python logging to be handled along with other framework logging that defaults to the `nifi-app.log` file appender.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcessLogReader.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        try {
+            String line = processReader.readLine();
+            while (line != null) {
+                if (line.length() > MINIMUM_LINE_LENGTH) {
+                    processLine(line);
+                }
+                line = processReader.readLine();
+            }
+        } catch (final IOException e) {
+            processLogger.error("Read Process output failed", e);
```
Comment (markap14) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcessLogReader.java:61]: The message here is quite vague; perhaps "Failed to read output of Python Process"?

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessLogReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            final BufferedReader reader = new BufferedReader(new StringReader(LINE_UNFORMATTED));
+            final Runnable command = new PythonProcessLogReader(reader);
+            command.run();
+        }
+
+        verify(logger).info(eq(LINE_UNFORMATTED));
+    }
+
+    @Test
+    void testMissingSeparator() {
```
Comment (markap14) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessLogReaderTest.java:101]: You mind adding a test for logging an empty string? I was wondering when reading the code above, if the `:` separator is the last character in the line, is that going to throw some sort of ArrayIndexOutOfBoundsException or something?

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcessLogReader.java (hunk 1 above)
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcessLogReader.java:61]: The logger name may help, but your suggestion is definitely clearer.

**Comment by exceptionfactory** [comment]
Thanks for the initial feedback @markap14! Based on further discussion, I will be following up with additional changes to incorporate log level updates from Logback.

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcessLogReader.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+     * Read lines from Process Reader and write log messages based on parsed level and named logger
+     */
+    @Override
+    public void run() {
+        final Queue<ParsedRecord> parsedRecords = new LinkedList<>();
+
+        try {
+            String line = processReader.readLine();
+            while (line != null) {
+                processLine(line, parsedRecords);
```
Comment (markap14) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcessLogReader.java:82]: I think within this `while` loop we should have a try/catch (Exception e)
Without it, if there's any Exception thrown from parsing, etc. the execution will escape the run() method without any sort of warning, and the logging will simply stop. Which will eventually cause the Python process' STDOUT to fill up and then the Python process will block.
Need to ensure that this thread never dies unexpectedly.

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcessLogReader.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+                if (parsedRecords.size() == 2 || !processReader.ready()) {
+                    final ParsedRecord parsedRecord = parsedRecords.remove();
+                    log(parsedRecord);
+                }
```
Comment (markap14) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcessLogReader.java:87]: I think the logic here is a bit off.
I created this Processor for testing:
```
from nifiapi.flowfiletransform import FlowFileTransform, FlowFileTransformResult
import logging

root_logger = logging.getLogger('')

class LogContents(FlowFileTransform):
    class Java:
        implements = ['org.apache.nifi.python.processor.FlowFileTransform']
    class ProcessorDetails:
        version = '0.0.1-SNAPSHOT'

    def __init__(self, **kwargs):
        pass

    def transform(self, context, flowFile):
        contents = flowFile.getContentsAsBytes().decode("utf-8")
        root_logger.info("Contents:\n1\n2\n3")
        root_logger.info(contents)
        return FlowFileTransformResult(relationship = "success")
```

I then sent in some FlowFiles. Each time, I get the log output of:
```
Contents
1
2
3
```

but not the output of the actual logger.
If I change the first log message instead to:
```
        root_logger.info("Contents:")
```

I get just the line "Contents:" on the first iteration. The second time the processor is run, it outputs the contents of the first FlowFile, followed by "Contents" again.

Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcessLogReader.java]: Thanks for the thorough feedback and testing. I made a change to check the readiness of the Process Reader separately, and then the example `LogContents` Processor writes the messages as expected.

**Comment by markap14** [comment]
Thanks @exceptionfactory all seems to be working well now. This is a huge improvement! +1 will merge to main
